Adds document for python coding standards (#120)
canonical-webteam/practices: coding/python.md

---
layout: base
---

# Python code standards

This document outlines the rules for writing Python code across all our Projects.

## Code style and formatting

> All Code should be written in accordance with [PEP 8](https://pep8.org/).

Reviewer: I think maybe it's worth being clear about which points here are absolute requirements and which are strong suggestions - maybe with a nod towards https://tools.ietf.org/html/rfc2119.

So, in this case I think we can say: 

```suggestion
All code must conform to [PEP 8](https://pep8.org/).
```

> Any pull requests that do no conform to this should be rejected by the CI.

Reviewer: ```suggestion

Project should be configured to block merging on pull requests that do not conform to these standards.
```

> To correctly format all your code please use [black](https://github.com/ambv/black). Since it differs in a small way from the standard **PEP8** make sure to pass `--line-length 79` when using it. The reason for black to do this is outlined [here](https://github.com/ambv/black#line-length).

Reviewer: I don't think we need to explain Black's choices here - it seems a bit off-topic. People can do that research themselves if they really want to.

We might want to add a note that CI should reject any pull requests that don't conform to these standard though.

## Testing

> All logic should be tested. If you add something new to the Codebase make sure that you add sufficient testing to ensure longterm maintainability of our projects.

Reviewer: Unfortunately I think this is just too far from either the reality of our existing codebases or our current practice to be taken seriously. Maybe we could say something like:

```suggestion
Although tests are lacking on many of our existing projects, writing tests is strongly encouraged. Please write tests for any new code if at all possible, and any efforts to improve test coverage on any project would be greatly appreciated.
```

## Documentation

> Document your functions with docstrings in accordance with [PEP 257](https://www.python.org/dev/peps/pep-0257/).

Reviewer: This is where I think https://tools.ietf.org/html/rfc2119 wording could be useful to make it clear whether this is a hard requirement or not. Again, our current practice falls quite a long way short of writing docstrings for *every* function, and I don't think that it will ever be 100% followed as a hard requirement. So, to be clear, I'd say:

```suggestion
You should document your functions with docstrings in accordance with [PEP 257](https://www.python.org/dev/peps/pep-0257/).
```

---
layout: base
---

# Python code standards

This document outlines the rules for writing Python code across all our Projects.

## Code style and formatting

All code must conform to [PEP 8](https://pep8.org/) and be formatted with [black](https://github.com/ambv/black) using the option `--line-length 79`.

Project should be configured to block merging on pull requests that do not conform to these standards.

## Testing

> Although tests are lacking on many of our existing projects, writing tests is strongly encouraged. Please write tests for any new code if at all possible, and any efforts to improve test coverage on any project would be greatly appreciated.

Reviewer 2: Only strongly encouraged? Shouldn't tests be mandatory?

Reviewer 2: > And unfortunately we're not in a place at the moment to enforce 100% test coverage on all new work.

Why's that?

Reviewer: I responded in possibly too much depth over at https://github.com/canonical-webteam/practices/issues/122#issuecomment-441273018.

I didn't want to start that discussion as part of this PR because I think it's better to initially document current practice, so we can build up the practices faster, but I'm very happy to explore evolving the policy.

## Code style and formatting

All code must conform to [PEP 8](https://pep8.org/) and be formatted with [black](https://github.com/ambv/black) using the option `--line-length 79`.

> Project should be configured to block merging on pull requests that do not conform to these standards.

Reviewer: I wasn't quite sure if you intended this line to be its own paragraph. If so, it needs a black line above it. But I don't mind either way.
